cosmwasm/packages/wasmi-runtime/src/wasm/query_chain.rs

```diff
@@ -0,0 +1,224 @@
+                                WasmEngineError::DeserializationError
+                            })?;
+
+                    Ok(Ok(Binary(decrypted)))
+                }
+                Err(err) => {
+                    error!(
+                                    "encrypt_and_query_chain() got an error while trying to decrypt the result for query {:?}, stopping wasm: {:?}",
+                                    String::from_utf8_lossy(&query),
+                                    err
+                                );
+
+                    return Err(WasmEngineError::DecryptionError);
+                }
+            }
+        }
+        Ok(Err(StdError::GenericErr { msg })) => match base64::decode(&msg) {
+            Err(err) => {
+                error!(
+                            "encrypt_and_query_chain() got an StdError as an answer {:?}, tried to decode the inner msg as bytes because it's encrypted, but got an error while trying to decode from base64. This usually means that the called contract panicked and the error is plaintext: {:?}",
+                            msg, err
+                        );
+                Ok(Err(StdError::GenericErr { msg }))
+            }
+            Ok(inner_error_bytes) => {
+                // query response returns without nonce and user_public_key appended to it
+                // because the sender is supposed to have them already
+                let inner_error_as_secret_msg = SecretMessage {
+                    nonce,
+                    user_public_key,
+                    msg: inner_error_bytes,
+                };
+
+                match inner_error_as_secret_msg.decrypt() {
+                            Err(err) => {
+                                error!(
+                                    "encrypt_and_query_chain() got an error while trying to decrypt the inner error for query {:?}, stopping wasm: {:?}",
+                                    String::from_utf8_lossy(&query),
+                                    err
+                                );
+
+                                return Err(WasmEngineError::DecryptionError);
+                            }
+                            Ok(decrypted) => {
+                                serde_json::from_slice(&decrypted).map_err(|err| {
+                                    error!("encrypt_and_query_chain() got an error while trying to deserialize the inner error as StdError: {:?}", err);
+                                    WasmEngineError::DeserializationError
+                                })?
+                            }
+                        }
+            }
+        },
+        Ok(Err(std_error)) => {
+            error!(
+                        "encrypt_and_query_chain() got an StdError as an answer, but it should be of type GenericErr and encrypted inside. Got instead: {:?}",
+                        std_error
+                    );
+            Ok(Err(std_error))
+        }
+    };
```

Comment:
Make a separate function(s). This entire thing is basically a `decrypt_answer` function, and inside you have smaller  functions, e.g. `decrypt_generic_err`. Would go a long way to making this easier to digest (you're returning stuff like ` Ok(Ok(Binary(decrypted)))` - this isn't easy to understand without the code being well structured)

Reply:
You're right. Will fix this once all tests pass.

The `Ok(Ok(...` stuff is part of CosmWasm's API for `Querier`.
The caller contract expects to receive `Result<Result<Binary, StdError>, SystemError>`.
- `Binary` - An encrypted answer from the callee.
- `StdError` - An encrypted error from the callee, or if panic a plaintext error.
- `SystemError` - A plaintext error from the VM of the callee. See [this](https://github.com/enigmampc/SecretNetwork/blob/7754e997655bb33d5d6f0614b280e4008a3630bd/cosmwasm/packages/wasmi-runtime/src/cosmwasm/system_error.rs#L8-L16) for more info.